sync/index.py

```diff
@@ -28,15 +28,15 @@
 
 
 class Index(metaclass=abc.ABCMeta):
-    name: Text = None
-    key_fields: Tuple[Text, ...] = None
+    name: Optional[Text] = None
+    key_fields: Tuple[Text, ...]
     unique = False
     value_cls: type = tuple
 
     # Overridden in subclasses using the constructor
     # This provides a kind of borg pattern where all instances of
     # the class have the same changes data
-    changes: Dict[Any, Any] = None
+    changes: Optional[Dict[Any, Any]] = None
```

Comment:
In this case the fact that we have to add `None` checks later on seems like a bug/limitation in mypy :/ We have code like
```py
target = self.changes
if target is None:
   return {}
# At this point mypy should know that `target` is Mapping[Any, Any]
for part in key:
    target = target[key]
    # Now it thinks that `target` is Optional[Mapping[Any, Any]] again,
    # when presumably it should be Any, although we could also work out the actual type
```
```